lab2/part1/myrouter_part1.py

```diff
@@ -0,0 +1,149 @@
+        else:
+            return None
+
+    def __setitem__(self, ip_addr, mac_addr):
+        self.add_mapping(ip_addr, mac_addr)
+        return
+
+    #Add or updates an IP to MAC address mapping
+    #  ip_addr      IPv4Address
+    #  mac_addr     EthAddr
+    def add_mapping(self, ip_addr, mac_addr):
+        if isinstance(ip_addr, IPv4Address) and isinstance(mac_addr, EthAddr):
+            self.map[ip_addr] = ARPContext.ARPEntry(mac_addr)
+        return
+
+    #Process an ARP requests in the provided packet. Return a reply packet
+    #if the target IP address is in our table
+    #   pkt         Arp packet to process
+    def arp_request(self, pkt: Arp):
+        reply_ip  = pkt.targetprotoaddr
+        reply_mac = self.map[reply_ip].mac_addr
+
+        #Construct reply packet
+        etp = Ethernet(ethertype = EtherType.ARP,
+                       src       = reply_mac,
+                       dst       = pkt.senderhwaddr)
+
+        arp = Arp(operation         = ArpOperation.Reply,
+                  senderhwaddr      = reply_mac,
+                  senderprotoaddr   = reply_ip,
+                  targethwaddr      = pkt.senderhwaddr,
+                  targetprotoaddr   = pkt.senderprotoaddr)
+
+        return_pkt = etp + arp
+        return return_pkt
+
+    #Process an ARP reply in the provided packet.
+    #   pkt         Arp packet to process
+    def arp_reply(self, pkt: Arp):
+        self.add_mapping(pkt.senderprotoaddr, pkt.senderhwaddr)
+
+    """
+    Class:          ARPContext.ArpEntry
+    Description:    Table entry pairing a timestamp with a MAC address
+    """
+    class ARPEntry:
+        def __init__(self, mac_addr, time=time.time()):
+            self.mac_addr   = mac_addr
+            self.timestamp  = time
+            return
+
+# end class ARPContext
+
+class Router(object):
+    def __init__(self, net):
+        self.net = net
+        self.local_proto_eth = ARPContext()     #Local address maps
+        self.other_proto_eth = ARPContext()     #Other address maps
+
+        #Cache IP->MAC mapping for local interfaces
+        for intf in net.interfaces():
+            self.local_proto_eth[intf.ipaddr] = intf.ethaddr
+        
+    #Main Router loop
+    def router_main(self):    
+        '''
+        Main method for router; we stay in a loop in this method, receiving
+        packets until the end of time.
+        '''
+        while True:
+            gotpkt = True
+            try:
+                _, input_port, pkt = self.net.recv_packet(timeout=1.0)
+            except NoPackets:
+                log_debug("No packets available in recv_packet")
+                gotpkt = False
+            except Shutdown:
+                log_debug("Got shutdown signal")
+                break
+
+            if gotpkt:
+                log_debug("Got a packet: {}".format(str(pkt)))
+
+                #Handle ARP
+                self.handle_arp(pkt.get_header(Arp), input_port)
```

Comment:
We should have the type of the packet here in a variable, and then decide (with an `if`) whether to call `handle_arp()` based on that type. And later in part 2 or 3 where there are extra packet types to handle, we can just add `elif` for each additional packet type that we must handle. We should also have a final `else` where we `log_debug` a message to indicate that an unsupported packet type was received.

Reply:
Refactored a bit and copied over the version of handle_arp from part2